tests/unittests/BackendTestUtils2.cpp

```diff
@@ -340,47 +342,60 @@ void trainConvNet(Tensor *inputs, Tensor *kernel1, Tensor *bias1,
   TC.learningRate = 0.03;
   TC.momentum = 0.3;
   TC.L2Decay = 0.01;
-  auto &mod = EE.getModule();
-  Function *F = mod.createFunction("main");
-  auto *var1 = createPlaceholder(mod, bindings, inputs, "var1");
-  auto *var2 = createPlaceholder(mod, bindings, selected, "var2");
-  auto *conv1 = F->createConv(bindings, "conv1", var1, 3, {5, 3}, {2, 1},
-                              {2, 1, 2, 1}, 1);
-  bindings.get(cast<Placeholder>(conv1->getFilter()))->assign(kernel1);
-  bindings.get(cast<Placeholder>(conv1->getBias()))->assign(bias1);
-  auto *reshape1 = F->createReshape("reshape1", conv1, shape1);
-  auto *conv2 = F->createConv(bindings, "conv2", reshape1, 2, 2, 2, 0, 1);
-  bindings.get(cast<Placeholder>(conv2->getFilter()))->assign(kernel2);
-  bindings.get(cast<Placeholder>(conv2->getBias()))->assign(bias2);
-  auto *reshape2 = F->createReshape("reshape2", conv2, shape2);
-  auto *softmax = F->createSoftMax("softmax", reshape2, var2);
-  auto *result = F->createSave("ret", softmax);
-  auto *resultTensor = bindings.allocate(result->getPlaceholder());
-
-  Function *TF = glow::differentiate(F, TC);
-  EE.compile(CompilationMode::Train, TF);
-
-  runBatch(EE, bindings, 8, sampleCounter, {var1, var2}, {inputs, selected});
-  EE.compile(CompilationMode::Infer, F);
-  updateInputPlaceholders(bindings, {var1, var2}, {inputs, selected});
-  EE.run(bindings);
-  out->assign(resultTensor);
+  Function *F;
+  Placeholder *var1, *var2;
+  for (auto *EE : engines) {
+    auto &mod = EE->getModule();
+    F = mod.createFunction("main");
+    var1 = createPlaceholder(mod, bindings, inputs, "var1");
+    var2 = createPlaceholder(mod, bindings, selected, "var2");
+    auto *conv1 = F->createConv(bindings, "conv1", var1, 3, {5, 3}, {2, 1},
+                                {2, 1, 2, 1}, 1);
+    bindings.get(cast<Placeholder>(conv1->getFilter()))->assign(kernel1);
+    bindings.get(cast<Placeholder>(conv1->getBias()))->assign(bias1);
+    auto *reshape1 = F->createReshape("reshape1", conv1, shape1);
+    auto *conv2 = F->createConv(bindings, "conv2", reshape1, 2, 2, 2, 0, 1);
+    bindings.get(cast<Placeholder>(conv2->getFilter()))->assign(kernel2);
+    bindings.get(cast<Placeholder>(conv2->getBias()))->assign(bias2);
+    auto *reshape2 = F->createReshape("reshape2", conv2, shape2);
+    auto *softmax = F->createSoftMax("softmax", reshape2, var2);
+    F->createSave("ret", softmax);
+  }
+  trainingBindings.allocate(EET.getModule().getPlaceholders());
+  inferBindings.allocate(EEI.getModule().getPlaceholders());
+  bindings.copyTrainedWeightsTo(trainingBindings);
```

Comment:
Hm, so this is necessary because we need to copy in e.g. the trainable PHs' backing tensors created by `createConv()` etc., since they've been initialized in `bindings`?

I guess the name `copyTrainedWeightsTo()` feels misleading/wrong now, because the PHs aren't actually trained yet. Perhaps it would be better to rename it to `copyTrainableWeightsTo()` since that's what's actually happening anyway.